src/compiler/checker.ts

```diff
@@ -15903,6 +15903,9 @@ namespace ts {
         }
 
         function distributeIndexOverObjectType(objectType: Type, indexType: Type, writing: boolean) {
+            if (shouldDeferIndexType(objectType)) {
+                return;
+            }
             // (T | U)[K] -> T[K] | U[K] (reading)
             // (T | U)[K] -> T[K] & U[K] (writing)
             // (T & U)[K] -> T[K] & U[K]

```

Comment:
Change the following line to:
```ts
if (objectType.flags & TypeFlags.Union || objectType.Flags & TypeFlags.Intersection && !shouldDeferIndexType(objectType)) ...
```

Reply:
I'm concerned that `shouldDeferIndexType` will also do (possibly expensive) work for union types, so I'd prefer only deferring for intersection types.